File: geodiff/tests/test_skip_tables.cpp
/*
 GEODIFF - MIT License
 Copyright (C) 2022 Alexander Bruy
*/

#include "gtest/gtest.h"
#include "geodiff_testutils.hpp"
#include "geodiff.h"
#include "geodiffutils.hpp"

TEST( SkipTablesSqlite3Test, test_skip_create )
{
  std::string testname( "test_skip_create" );
  std::string base = pathjoin( testdir(), "skip_tables", "base.gpkg" );
  std::string modified_all = pathjoin( testdir(), "skip_tables", "modified_all.gpkg" );
  std::string modified_points = pathjoin( testdir(), "skip_tables", "modified_points.gpkg" );
  std::string changeset_points = pathjoin( tmpdir(), testname, "changeset_points.bin" );
  std::string patched_points = pathjoin( tmpdir(), testname, "patched_points.gpkg" );

  makedir( pathjoin( tmpdir(), testname ) );

  // ignore lines table when creating changeset
  std::string tablesToSkip( "lines" );
  Context *ctx = static_cast<Context *>( testContext() );
  ctx->setTablesToSkip( tablesToSkip );

  int res = GEODIFF_createChangeset( testContext(), base.c_str(), modified_all.c_str(), changeset_points.c_str() );
  EXPECT_EQ( res, GEODIFF_SUCCESS );

  int nchanges = GEODIFF_changesCount( testContext(), changeset_points.c_str() );
  EXPECT_EQ( nchanges, 4 );

  // reconstuct changes in the points layer
  filecopy( patched_points, base );
  res = GEODIFF_applyChangeset( testContext(), patched_points.c_str(), changeset_points.c_str() );
  EXPECT_EQ( res, GEODIFF_SUCCESS );

  // check that now it is same file with modified
  EXPECT_TRUE( equals( patched_points, modified_points, false ) );

  // reset skip list
  tablesToSkip = "";
  ctx->setTablesToSkip( tablesToSkip );
}

TEST( SkipTablesSqlite3Test, test_skip_apply )
{
  std::string testname( "test_skip_apply" );
  std::string base = pathjoin( testdir(), "skip_tables", "base.gpkg" );
  std::string modified_all = pathjoin( testdir(), "skip_tables", "modified_all.gpkg" );
  std::string modified_points = pathjoin( testdir(), "skip_tables", "modified_points.gpkg" );
  std::string changeset = pathjoin( tmpdir(), testname, "changeset.bin" );
  std::string patched_points = pathjoin( tmpdir(), testname, "patched_points.gpkg" );

  makedir( pathjoin( tmpdir(), testname ) );

  int res = GEODIFF_createChangeset( testContext(), base.c_str(), modified_all.c_str(), changeset.c_str() );
  EXPECT_EQ( res, GEODIFF_SUCCESS );

  int nchanges = GEODIFF_changesCount( testContext(), changeset.c_str() );
  EXPECT_EQ( nchanges, 6 );

  // ignore lines table when applying changeset
  std::string tablesToSkip( "lines" );
  Context *ctx = static_cast<Context *>( testContext() );
  ctx->setTablesToSkip( tablesToSkip );

  // reconstuct changes in the points layer
  filecopy( patched_points, base );
  res = GEODIFF_applyChangeset( testContext(), patched_points.c_str(), changeset.c_str() );
  EXPECT_EQ( res, GEODIFF_SUCCESS );

  // check that now it is same file with modified
  EXPECT_TRUE( equals( patched_points, modified_points, false ) );

  // reset skip list
  tablesToSkip = "";
  ctx->setTablesToSkip( tablesToSkip );
}

int main( int argc, char **argv )
{
  testing::InitGoogleTest( &argc, argv );
  init_test();
  int ret =  RUN_ALL_TESTS();
  finalize_test();
  return ret;
}

File: pygeodiff/tests/test_skip_tables.py
# -*- coding: utf-8 -*-
"""
    :copyright: (c) 2022 Alexander Bruy
    :license: MIT, see LICENSE for more details.
"""

from .testutils import *
import os
import shutil
import pygeodiff

class UnitTestsPythonSingleCommit(GeoDiffTests):
    def test_skip_create(self):
        base = geodiff_test_dir() + "/" + "skip_tables" + "/" + "base.gpkg"
        modified = geodiff_test_dir() + "/" + "skip_tables" + "/" + "modified_all.gpkg"
        modified_points = geodiff_test_dir() + "/" + "skip_tables" + "/" + "modified_points.gpkg"
        changeset = tmpdir() + "/py" + "test_skip_create" + "/" + "changeset_points.bin"
        changeset2 = tmpdir() + "/py" + "test_skip_create" + "/" + "changeset_points2.bin"
        changeset_inv = tmpdir() + "/py" + "test_skip_create" + "/" + "changeset_inv.bin"
        patched = tmpdir() + "/py" + "test_skip_create" + "/" + "patched_points.gpkg"

        create_dir("test_skip_create")

        # ignore lines table when creating changeset
        self.geodiff.set_tables_to_skip(["lines"])

        # create changeset
        self.geodiff.create_changeset(base, modified, changeset)
        check_nchanges(self.geodiff, changeset, 4)

        # apply changeset
        shutil.copyfile(base, patched)
        self.geodiff.apply_changeset(patched, changeset)

        # check that now it is same file
        self.geodiff.create_changeset(patched, modified, changeset2)
        check_nchanges(self.geodiff, changeset2, 0)

        # check we can create inverted changeset
        os.remove(changeset2)
        self.geodiff.invert_changeset(changeset, changeset_inv)
        self.geodiff.apply_changeset(patched, changeset_inv)
        self.geodiff.create_changeset_dr("sqlite", "", patched, "sqlite", "", base, changeset2)
        check_nchanges(self.geodiff, changeset2, 0 )

        self.geodiff.set_tables_to_skip([])

    def test_skip_apply(self):
        base = geodiff_test_dir() + "/" + "skip_tables" + "/" + "base.gpkg"
        modified = geodiff_test_dir() + "/" + "skip_tables" + "/" + "modified_all.gpkg"
        modified_points = geodiff_test_dir() + "/" + "skip_tables" + "/" + "modified_points.gpkg"
        changeset = tmpdir() + "/py" + "test_skip_apply" + "/" + "changeset_points.bin"
        changeset2 = tmpdir() + "/py" + "test_skip_apply" + "/" + "changeset_points2.bin"
        changeset_inv = tmpdir() + "/py" + "test_skip_apply" + "/" + "changeset_inv.bin"
        patched = tmpdir() + "/py" + "test_skip_apply" + "/" + "patched_points.gpkg"

        create_dir("test_skip_apply")

        # create changeset
        self.geodiff.create_changeset(base, modified, changeset)
        check_nchanges(self.geodiff, changeset, 6)

        # ignore lines table when creating changeset
        self.geodiff.set_tables_to_skip(["lines"])

        # apply changeset
        shutil.copyfile(base, patched)
        self.geodiff.apply_changeset(patched, changeset)

        # check that now it is same file
        self.geodiff.create_changeset(patched, modified, changeset2)
        check_nchanges(self.geodiff, changeset2, 0)

        # check we can create inverted changeset
        os.remove(changeset2)
        self.geodiff.invert_changeset(changeset, changeset_inv)
        self.geodiff.apply_changeset(patched, changeset_inv)
        self.geodiff.create_changeset_dr("sqlite", "", patched, "sqlite", "", base, changeset2)
        check_nchanges(self.geodiff, changeset2, 0 )

        self.geodiff.set_tables_to_skip([])
